<a href="https://colab.research.google.com/github/gabrwell/analise_de_dados/blob/main/Checkpoint2_SERS.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Appliance Energy Prediction**

In [17]:
#Importando Bibliotecas
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.ensemble import RandomForestRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import numpy as np

In [18]:
# colocando o dataset e pritando as informações dele
df = pd.read_csv('/content/energydata_complete.csv', on_bad_lines='skip')

print(df.head())
print(df.info())
print(df.describe())

                  date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility

In [19]:
X = df.drop(['date', 'Appliances'], axis=1)
y = df['Appliances']

In [20]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [21]:
# Regressão Linear
lr = LinearRegression()
lr.fit(X_train, y_train)
y_pred_lr = lr.predict(X_test)

In [22]:
#Árvore de Regressão (Decision Tree)
dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)

In [23]:
#Random Forest
rf = RandomForestRegressor(random_state=42)
rf.fit(X_train, y_train)
y_pred_rf = rf.predict(X_test)

In [24]:
# Avaliação dos modelos usando R², RMSE e MAE
def print_metrics(y_true, y_pred, model_name):
    r2 = r2_score(y_true, y_pred)
    rmse = np.sqrt(mean_squared_error(y_true, y_pred))
    mae = mean_absolute_error(y_true, y_pred)
    print(f"\n{model_name}:")
    print(f"R² = {r2:.4f}")
    print(f"RMSE = {rmse:.2f}")
    print(f"MAE = {mae:.2f}")

In [25]:
print_metrics(y_test, y_pred_lr, "Regressão Linear")
print_metrics(y_test, y_pred_dt, "Árvore de Decisão")
print_metrics(y_test, y_pred_rf, "Random Forest")


Regressão Linear:
R² = 0.1693
RMSE = 91.17
MAE = 52.55

Árvore de Decisão:
R² = 0.1798
RMSE = 90.60
MAE = 38.87

Random Forest:
R² = 0.5322
RMSE = 68.42
MAE = 32.90


Conclusão:

- **Random Forest** tem o maior valor de R² (0.5322), indicando que explica uma proporção maior da variância no consumo de energia em comparação com a Regressão Linear (0.1693) e a Árvore de Decisão (0.1798).
- **Random Forest** também tem os menores valores de RMSE (68.42) e MAE (32.90), o que significa que suas previsões estão, em média, mais próximas dos valores reais do que os outros dois modelos.

Portanto, o modelo **Random Forest** é o melhor entre os três para explicar e prever o consumo de energia com base nesta análise.

# **Smart Grid Stability**

**Reasoning**:
The previous command failed because the pandas library was not imported in the current session. I need to re-import pandas before loading the data.



In [26]:
import pandas as pd

df_stability = pd.read_csv('/content/smart_grid_stability_augmented.csv')

print(df_stability.head())
print(df_stability.info())
print(df_stability.describe())

       tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Data columns (total 14 column

In [27]:
X_stability = df_stability.drop(['stab', 'stabf'], axis=1)
y_stability = df_stability['stabf']

In [28]:
from sklearn.model_selection import train_test_split

X_train_stability, X_test_stability, y_train_stability, y_test_stability = train_test_split(
    X_stability, y_stability, test_size=0.2, random_state=42
)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

rf_stability = RandomForestClassifier(random_state=42)
rf_stability.fit(X_train_stability, y_train_stability)

In [ ]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, confusion_matrix

y_pred_stability = rf_stability.predict(X_test_stability)

accuracy = accuracy_score(y_test_stability, y_pred_stability)
precision = precision_score(y_test_stability, y_pred_stability, pos_label='stable')
recall = recall_score(y_test_stability, y_pred_stability, pos_label='stable')
f1 = f1_score(y_test_stability, y_pred_stability, pos_label='stable')
conf_matrix = confusion_matrix(y_test_stability, y_pred_stability, labels=['stable', 'unstable'])

print(f"Accuracy: {accuracy:.4f}")
print(f"Precision (Stable): {precision:.4f}")
print(f"Recall (Stable): {recall:.4f}")
print(f"F1-score (Stable): {f1:.4f}")
print("\nConfusion Matrix:")
print(conf_matrix)

In [ ]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, f1_score, confusion_matrix
import numpy as np

In [ ]:
# Decision Tree Classifier
dt_stability = DecisionTreeClassifier(random_state=42)
dt_stability.fit(X_train_stability, y_train_stability)
y_pred_dt_stability = dt_stability.predict(X_test_stability)

# KNN Classifier
knn_stability = KNeighborsClassifier()
knn_stability.fit(X_train_stability, y_train_stability)
y_pred_knn_stability = knn_stability.predict(X_test_stability)

# Logistic Regression
lr_stability = LogisticRegression(random_state=42)
lr_stability.fit(X_train_stability, y_train_stability)
y_pred_lr_stability = lr_stability.predict(X_test_stability)

In [ ]:
def print_classification_metrics(y_true, y_pred, model_name):
    accuracy = accuracy_score(y_true, y_pred)
    f1 = f1_score(y_true, y_pred, pos_label='stable')
    conf_matrix = confusion_matrix(y_true, y_pred, labels=['stable', 'unstable'])

    print(f"\n{model_name}:")
    print(f"Accuracy = {accuracy:.4f}")
    print(f"F1-score (Stable) = {f1:.4f}")
    print("\nConfusion Matrix:")
    print(conf_matrix)

print_classification_metrics(y_test_stability, y_pred_dt_stability, "Decision Tree Classifier")
print_classification_metrics(y_test_stability, y_pred_knn_stability, "KNN Classifier")
print_classification_metrics(y_test_stability, y_pred_lr_stability, "Logistic Regression")